pkg/webhook/topology/mutator.go

```diff
@@ -0,0 +1,115 @@
+//  Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package topology
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	gardenContext "github.com/gardener/gardener/extensions/pkg/webhook/context"
+	"github.com/gardener/gardener/pkg/extensions"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-azure/pkg/azure"
+)
+
+type mutator struct {
+	client client.Client
+	log    logr.Logger
+}
+
+// New initializes a new mutator.
+func New(logger logr.Logger) *mutator {
+	return &mutator{
+		log: logger,
+	}
+}
+
+func (m *mutator) InjectClient(client client.Client) error {
+	m.client = client
+	return nil
+}
+
+func (m *mutator) Mutate(ctx context.Context, new, _ client.Object) error {
+	if new == nil {
+		return nil
+	}
+
+	// do not try to mutate pods that are getting deleted
+	if new.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	newPod, ok := new.(*v1.Pod)
+	if !ok {
+		return fmt.Errorf("could not mutate: object is not type Pod")
+	}
+
+	gctx := gardenContext.NewGardenContext(m.client, new)
+	cluster, err := gctx.GetCluster(ctx)
+	if err != nil {
+		m.log.Error(err, "failed to mutate resource: %s/%s", new.GetNamespace(), new.GetName())
```

Comment:
This is no printf style logger, please use named fields instead. You can use the following example:

```
log := m.log.Error("pod", client.ObjectKeyFromObject(newPod))
...
log.Error(err, "failed to mutate resource")
```

Reply:
Thank you for pointing this out. I proceeded differently and remove the logging because the webhook handler itself logs the errors so there is no need to do it a second time IMO.